### PR TITLE
Improve Psychoacoustic model and some ASR maintenance

### LINKED ISSUE #929: Psychoacoustic model results in unstable imperceptible loss
The current Psychoacoustic model of the framework-agnostic imperceptible ASR attack results in an unstable imperceptible loss. This is due to very small differences in the different STFT implementations of `librosa` and `scipy`. It turns out that both STFT's are identical only up to 5 decimals. After switching to `librosa`'s STFT, the imperceptible loss stabilizes. Thus this difference is not negligible at all:
```
np.testing.assert_almost_equal(np.abs(stft_stft), np.abs(stft_librosa), decimal=6)
*** AssertionError:
Arrays are not almost equal to 6 decimals
Mismatched elements: 351 / 290075 (0.121%)
Max absolute difference: 7.6293945e-06
Max relative difference: 0.00265919
```

**System information (please complete the following information):**
 - Fedora 33
 - Python 3.6.10
 - ART 1.5

### PR DESCRIPTION
**Pull request by hesseltuinhof**
# Description

This PR fixes the Psychoacoustic model and stabilizes the imperceptible loss. Especially, switching to `librosa`'s STFT and using scalar PSD maximum, fixes stability issues with the imperceptible loss.

Fixes #929

Furthermore, the PR includes several minor improvements to ensure that framework-agnostic imperceptible ASR works as expected. These include:

* Return transcription as default in predict as explained in docstring  (Deepspeech estimator)
* Fix wrong gradient shape if `batch_size=1` as explained in docstring (Deepspeech estimator)
* Use periodic window for STFT instead symmetric window (Deepspeech-specific attack)
* Ensure correct `dtype` and streamline formulas to increase numerical stability (Psychoacoustic models)
* Refactor loss theta to only calculate masking thresholds once (Framework-agnostic attack)
* Add early stopping if loss theta < 0.05 to avoid running into gradients with `nan` values (Framework-agnostic attack)

## Type of change

Please check all relevant options.

- [x] Improvement (non-breaking)
- [x] Bug fix (non-breaking)

# Testing

Changes have been tested against the ten LibriSpeech test samples provided by Qin et al.

**Test Configuration**:
- Fedora 33
- Python 3.6.10
- ART 1.5.3 development

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/attacks/evasion/imperceptible_asr/imperceptible_asr.py [hunk 1]
```diff
@@ -76,6 +76,7 @@ def __init__(
         learning_rate_2: float = 1.0,
         max_iter_2: int = 4000,
         batch_size: int = 16,
+        **kwargs,
```
Comment (beat-buesser) [inline art/attacks/evasion/imperceptible_asr/imperceptible_asr.py:79]: Why do you add `kwargs`? Can we add the argument `loss_theta_min` instead?

File: art/attacks/evasion/imperceptible_asr/imperceptible_asr.py [hunk 2]
```diff
@@ -292,8 +318,28 @@ def _create_imperceptible(self, x: np.ndarray, x_adversarial: np.ndarray, y: np.
 
         return np.array(x_imperceptible, dtype=object)
 
+    def _stabilized_threshold_and_psd_maximum(self, x: np.ndarray):
```
Comment (beat-buesser) [inline art/attacks/evasion/imperceptible_asr/imperceptible_asr.py:321]: What is the type of the returned value?

File: art/attacks/evasion/imperceptible_asr/imperceptible_asr.py [hunk 3]
```diff
@@ -617,8 +648,11 @@ def absolute_threshold_hearing(self) -> np.ndarray:
             valid_domain = np.logical_and(20 <= self.fft_frequencies, self.fft_frequencies <= 2e4)
             freq = self.fft_frequencies[valid_domain] * 0.001
 
-            # outside valid ATH domain, set values to np.inf
-            self._absolute_threshold_hearing = np.ones(valid_domain.shape) * np.inf
+            # outside valid ATH domain, set values to -np.inf
+            # note: This ensures that every possible masker in the bins <=20Hz is valid. As a consequence, the global
+            #   masking threshold formula will always return a value different to np.inf
```
Comment (beat-buesser) [inline art/attacks/evasion/imperceptible_asr/imperceptible_asr.py:653]:
```suggestion
            # masking threshold formula will always return a value different to np.inf
```

File: art/attacks/evasion/imperceptible_asr/imperceptible_asr.py [hunk 4]
```diff
@@ -632,29 +666,25 @@ def power_spectral_density(self, audio: np.ndarray) -> Tuple[np.ndarray, np.ndar
         :return: PSD matrix of shape `(window_size // 2 + 1, frame_length)` and maximum vector of shape
         `(frame_length)`.
         """
+        import librosa
```
Comment (beat-buesser) [inline art/attacks/evasion/imperceptible_asr/imperceptible_asr.py:669]: Should we add `librosa` to `requirements.txt`?

File: art/attacks/evasion/imperceptible_asr/imperceptible_asr.py (hunk 2 above)
Comment (beat-buesser) [reply art/attacks/evasion/imperceptible_asr/imperceptible_asr.py:321]: Please add in the docstring a description of argument `x`.

**Review by beat-buesser — CHANGES_REQUESTED** [review]
Hi @hesseltuinhof Looks good to me! I have only a few minor update requests.

File: art/attacks/evasion/imperceptible_asr/imperceptible_asr.py [hunk 5]
```diff
@@ -102,6 +105,7 @@ def __init__(
         self.max_iter_2 = max_iter_2
         self._targeted = True
         self.batch_size = batch_size
+        self._loss_theta_min = loss_theta_min
```
Comment (beat-buesser) [inline art/attacks/evasion/imperceptible_asr/imperceptible_asr.py:108]:
```suggestion
        self.loss_theta_min = loss_theta_min
```

File: art/attacks/evasion/imperceptible_asr/imperceptible_asr.py [hunk 6]
```diff
@@ -284,16 +306,53 @@ def _create_imperceptible(self, x: np.ndarray, x_adversarial: np.ndarray, y: np.
                         alpha[j] = max(alpha[j] * 0.8, alpha_min)
                 logger.info("Current iteration %s, alpha %s, loss theta %s", i, alpha, loss_theta)
 
+            # note: avoids nan values in loss theta, which can occur when loss converges to zero.
+            for j in range(batch_size):
+                if loss_theta[j] < self._loss_theta_min and not early_stop[j]:
+                    logger.warning(
+                        "Batch sample %s reached minimum threshold of %s for theta loss.", j, self._loss_theta_min
```
Comment (beat-buesser) [inline art/attacks/evasion/imperceptible_asr/imperceptible_asr.py:313]:
```suggestion
                if loss_theta[j] < self.loss_theta_min and not early_stop[j]:
                    logger.warning(
                        "Batch sample %s reached minimum threshold of %s for theta loss.", j, self.loss_theta_min
```

File: art/attacks/evasion/imperceptible_asr/imperceptible_asr.py [hunk 7]
```diff
@@ -284,16 +306,53 @@ def _create_imperceptible(self, x: np.ndarray, x_adversarial: np.ndarray, y: np.
...
         for j in range(batch_size):
             if x_imperceptible[j] is None:
                 logger.critical("Adversarial attack stage 2 for x_%s was not successful", j)
                 x_imperceptible[j] = x_perturbed[j]
 
-        return np.array(x_imperceptible, dtype=object)
+        return np.array(x_imperceptible, dtype=dtype)
+
+    def _stabilized_threshold_and_psd_maximum(self, x: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
+        """Return batch of stabilized masking thresholds and PSD maxima.
```
Comment (beat-buesser) [inline art/attacks/evasion/imperceptible_asr/imperceptible_asr.py:331]:
```suggestion
        """
        Return batch of stabilized masking thresholds and PSD maxima.
```

File: art/attacks/evasion/imperceptible_asr/imperceptible_asr.py [hunk 8]
```diff
@@ -102,6 +105,7 @@ def __init__(
         self.max_iter_2 = max_iter_2
         self._targeted = True
```
Comment (minhitbk) [inline art/attacks/evasion/imperceptible_asr/imperceptible_asr.py:106]:
```suggestion
        self.targeted = True
```

File: art/attacks/evasion/imperceptible_asr/imperceptible_asr.py [hunk 9]
```diff
@@ -158,7 +160,10 @@ def generate(self, x: np.ndarray, y: np.ndarray, **kwargs) -> np.ndarray:
 
             # create batch of adversarial examples
             x_imperceptible[begin:end] = self._generate_batch(x[begin:end], y[begin:end])
-        return np.array(x_imperceptible, dtype=object)
+
+        # for ragged input, use np.object dtype
+        dtype = np.float32 if x.ndim != 1 else np.object
```
Comment (minhitbk) [inline art/attacks/evasion/imperceptible_asr/imperceptible_asr.py:165]:
```suggestion
        dtype = ART_NUMPY_DTYPE if x.ndim != 1 else np.object
```

File: art/attacks/evasion/imperceptible_asr/imperceptible_asr.py [hunk 10]
```diff
@@ -190,19 +197,22 @@ def _create_adversarial(self, x, y) -> np.ndarray:
         """
         batch_size = x.shape[0]
 
+        # for ragged input, use np.object dtype
+        dtype = np.float32 if x.ndim != 1 else np.object
```
Comment (minhitbk) [inline art/attacks/evasion/imperceptible_asr/imperceptible_asr.py:201]:
```suggestion
        dtype = ART_NUMPY_DTYPE if x.ndim != 1 else np.object
```

File: art/attacks/evasion/imperceptible_asr/imperceptible_asr.py [hunk 11]
```diff
@@ -246,22 +256,34 @@ def _create_imperceptible(self, x: np.ndarray, x_adversarial: np.ndarray, y: np.
         batch_size = x.shape[0]
         alpha_min = 0.0005
 
-        alpha = np.array([self.alpha] * batch_size)
+        # for ragged input, use np.object dtype
+        dtype = np.float32 if x.ndim != 1 else np.object
```
Comment (minhitbk) [inline art/attacks/evasion/imperceptible_asr/imperceptible_asr.py:260]:
```suggestion
        dtype = ART_NUMPY_DTYPE if x.ndim != 1 else np.object
```

File: art/attacks/evasion/imperceptible_asr/imperceptible_asr.py [hunk 12]
```diff
@@ -246,22 +256,34 @@ def _create_imperceptible(self, x: np.ndarray, x_adversarial: np.ndarray, y: np.
...
+
+        early_stop = [False] * batch_size
+
+        alpha = np.array([self.alpha] * batch_size, dtype=np.float32)
```
Comment (minhitbk) [inline art/attacks/evasion/imperceptible_asr/imperceptible_asr.py:264]:
```suggestion
        alpha = np.array([self.alpha] * batch_size, dtype=ART_NUMPY_DTYPE)
```

File: art/attacks/evasion/imperceptible_asr/imperceptible_asr.py [hunk 13]
```diff
@@ -348,7 +394,9 @@ def _loss_gradient_masking_threshold(
             gradient = gradient_padded[:length]
             gradients.append(gradient)
 
-        return np.array(gradients, dtype=object), loss
+        # for ragged input, use np.object dtype
+        dtype = np.float32 if x.ndim != 1 else np.object
```
Comment (minhitbk) [inline art/attacks/evasion/imperceptible_asr/imperceptible_asr.py:398]:
```suggestion
        dtype = ART_NUMPY_DTYPE if x.ndim != 1 else np.object
```

File: art/attacks/evasion/imperceptible_asr/imperceptible_asr.py [hunk 14]
```diff
@@ -632,29 +686,27 @@ def power_spectral_density(self, audio: np.ndarray) -> Tuple[np.ndarray, np.ndar
         :return: PSD matrix of shape `(window_size // 2 + 1, frame_length)` and maximum vector of shape
         `(frame_length)`.
         """
+        import librosa
+
         # compute short-time Fourier transform (STFT)
+        audio_float = audio.astype(np.float32)
```
Comment (minhitbk) [inline art/attacks/evasion/imperceptible_asr/imperceptible_asr.py:692]:
```suggestion
        audio_float = audio.astype(ART_NUMPY_DTYPE)
```

File: art/estimators/speech_recognition/tensorflow_lingvo.py [hunk 15]
```diff
@@ -513,7 +513,9 @@ def _loss_gradient_per_batch(self, x: np.ndarray, y: np.ndarray) -> np.ndarray:
             gradient = gradient_padded[:length]
             gradients.append(gradient)
 
-        return np.array(gradients, dtype=object)
+        # for ragged input, use np.object dtype
+        dtype = np.float32 if x.ndim != 1 else np.object
```
Comment (minhitbk) [inline art/estimators/speech_recognition/tensorflow_lingvo.py:517]:
```suggestion
        dtype = ART_NUMPY_DTYPE if x.ndim != 1 else np.object
```

File: art/estimators/speech_recognition/tensorflow_lingvo.py [hunk 16]
```diff
@@ -539,7 +541,9 @@ def _loss_gradient_per_sequence(self, x: np.ndarray, y: np.ndarray) -> np.ndarra
             gradient = self._sess.run(self._loss_gradient_op, feed_dict)
             gradients.append(np.squeeze(gradient))
 
-        return np.array(gradients, dtype=object)
+        # for ragged input, use np.object dtype
+        dtype = np.float32 if x.ndim != 1 else np.object
```
Comment (minhitbk) [inline art/estimators/speech_recognition/tensorflow_lingvo.py:545]:
```suggestion
        dtype = ART_NUMPY_DTYPE if x.ndim != 1 else np.object
```

**Review by minhitbk — APPROVED** [review]
Hi @hesseltuinhof : I have just a few minor comments. One of that is that I suggest to import ART_NUMPY_DTYPE and use it to replace any `np.float32` in your code. Similarly, you can define one for tf.float32. I remember torch.float32 is also defined somewhere which we can confirm with @beat-buesser.

**Comment by hesseltuinhof** [comment]
Thanks for the reviews. All issues have been addressed. The choice for using `np.float32` is deliberate as discussed.

**Review by beat-buesser — APPROVED** [review]

**Comment by beat-buesser** [comment]
Hi @hesseltuinhof Thank you very much! These updates make the Imperceptible ASR attack even stronger!
